lib/libc/sys/ioctl.c

```diff
@@ -88,9 +89,13 @@ _ioctl(int fd, unsigned long com, ...)
 		} else {
 			data = va_arg(ap, void *);
 		}
-		va_end(ap);
 	} else {
-		data = NULL;
+		if (cheri_getlen((void*)ap) == 16)
+			data = va_arg(ap, void *);
+		else
+			data = NULL;
```

Comment:
I think we want to handle the sane cases.  I believe that's int (char is promoted to int IIRC), long, and void *. I've chosen int##_t because I believe we want sign extension here but I could be wrong.

```suggestion
#ifdef __CHERI_PURE_CAPABILITY__
		if (cheri_getlen((void*)ap) == sizeof(void *))
			data = va_arg(ap, void *);
		else if (cheri_getlen((void*)ap) == sizeof(uint64_t))
			data = (void *)(intptr_t)va_arg(ap,  int64_t);
		else if (cheri_getlen((void*)ap) == sizeof(int32_t))
			data = (void *)(intptr_t)va_arg(ap, int32_t);
		else
			data = NULL;
#else
		/* Assume this was passed in a register. */
		data = (void *)(intptr_t)va_arg(ap, register_t);
#endif
```